tests/test_trainer.py

```diff
@@ -0,0 +1,69 @@
+import shutil
```

Comment:
Before this PR, tests ran in 4s, now they need 13s according to github actions run logs. 
Any chance of speeding this up?

Reply:
It's expected as now we initialize the whole trainer class (which is huge), which includes a lot of stuff happening. Also we do two passes of data (one in trainer tests, one in data tests). 

The first tests were to short as they were incomplete.